crates/storage/trie_db/layering.rs

```diff
@@ -1,181 +1,88 @@
-        qfilter::Filter::new_resizeable(1_000_000, 100_000_000, 0.02)
-            .inspect_err(|e| tracing::warn!("could not create trie layering bloom filter {e}"))
+    /// Obtain the cached value from any layer given its key.
+    pub fn get(&self, key: &[u8]) -> Option<&[u8]> {
+        self.data.get(key).map(|entry| &*entry.value)
     }
 
-    pub fn get(&self, state_root: H256, key: &[u8]) -> Option<Vec<u8>> {
-        // Fast check to know if any layer may contains the given key.
-        // We can only be certain it doesn't exist, but if it returns true it may or not exist (false positive).
-        if let Some(filter) = &self.bloom
-            && !filter.contains(key)
-        {
-            // TrieWrapper goes to db when returning None.
-            return None;
-        }
-
-        let mut current_state_root = state_root;
+    // /// Write a batch of items into the cache at the last layer.
+    // ///
+    // /// Items that were already present in that layer will be overwritten. Use
+    // /// [`TrieLayerCache::commit`] to advance the layers before putting items into the new layer.
+    // pub fn put_iter(&mut self, iter: impl IntoIterator<Item = (Arc<[u8]>, Arc<[u8]>)>) {
+    // }
 
-        while let Some(layer) = self.layers.get(&current_state_root) {
-            if let Some(value) = layer.nodes.get(key) {
-                return Some(value.clone());
-            }
-            current_state_root = layer.parent;
-            if current_state_root == state_root {
-                // TODO: check if this is possible in practice
-                // This can't happen in L1, due to system contracts irreversibly modifying state
-                // at each block.
-                // On L2, if no transactions are included in a block, the state root remains the same,
-                // but we handle that case in put_batch. It may happen, however, if someone modifies
-                // state with a privileged tx and later reverts it (since it doesn't update nonce).
-                panic!("State cycle found");
+    /// Return an iterator to extract the elements of the 128th layer.
+    ///
+    /// If there are not yet 128 layers, it'll return an empty vec.
+    pub fn commit_and_put_iter(
+        self: &mut Arc<Self>,
+        iter: impl IntoIterator<Item = (Arc<[u8]>, Arc<[u8]>)>,
+    ) -> Vec<(Arc<[u8]>, Arc<[u8]>)> {
+        let self_mut = Arc::make_mut(self);
+
+        // Commit last layer.
+        let mut items = Vec::new();
+        self_mut.data.retain(|key, entry| {
+            entry.layers <<= 1;
+            if entry.layers == 0 {
+                items.push((
+                    Arc::clone(key),
+                    if entry.previous.is_empty() {
+                        mem::take(&mut entry.value)
+                    } else {
+                        entry.previous.remove(0)
+                    },
+                ));
+
+                true
+            } else {
+                false
```

Comment:
I think these `true` and `false` should be the other way around: we want to retain the items that are not in last layer